src/common/ult_stack_mmap.c

```diff
@@ -0,0 +1,612 @@
+/**
+ * (C) Copyright 2021-2024 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * \file
+ *
+ * This file implements an alternate/external way for ULTs stacks allocation.
+ * It is based on mmap() of MAP_STACK|MAP_GROWSDOWN regions, in order to
+ * allow overrun detection along with automatic growth capability.
+ */
+
+#define D_LOGFAC DD_FAC(stack)
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include <errno.h>
+#include <string.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <search.h>
+#include <sys/mman.h>
+
+#include <daos/common.h>
+#include <daos_srv/daos_engine.h>
+#include <daos/daos_abt.h>
+#include <daos/ult_stack_mmap.h>
+
+#define MMAP_ULT_STACK_PROT     PROT_READ | PROT_WRITE
+#define MMAP_ULT_STACK_FLAGS    MAP_PRIVATE | MAP_ANONYMOUS | MAP_NORESERVE
+#define MMAP_ULT_STACK_FD       -1
+#define MMAP_ULT_STACK_OFFSET   0
+
+#define MREMAP_ULT_STACK_PROT   PROT_READ | PROT_WRITE
+#define MREMAP_ULT_STACK_FLAGS  MAP_PRIVATE | MAP_ANONYMOUS | MAP_STACK | MAP_FIXED | MAP_NORESERVE
+#define MREMAP_ULT_STACK_FD     -1
+#define MREMAP_ULT_STACK_OFFSET 0
+
+#define MMAP_GUARD_PAGE_PROT    PROT_NONE
+#define MMAP_GUARD_PAGE_FLAGS   MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED | MAP_NORESERVE
+#define MMAP_GUARD_PAGE_FD      -1
+#define MMAP_GUARD_PAGE_OFFSET  0
+
+/** Methods of ULT thread creation */
+enum thread_create_flags {
+	/** Create a new ULT and schedule it in a given pool */
+	TCF_ON_POOL = (1 << 0),
+	/** Create a new ULT associated with an execution stream */
+	TCF_ON_XSTREAM = (1 << 1)
+};
+
+/** Pool of mmap ULT free stack */
+struct stack_mmap_pool {
+	/** Binary tree holding list of ULT stacks */
+	void      *smp_rec_btree;
+	/** Lock protecting the binary tree smp_rec_btree */
+	ABT_rwlock smp_rec_btree_rwlock;
```

Comment:
Why can't we make it per xstream? I'm afraid that the performance could be impact by the locking contention significantly, the overhead of ABT lock creation isn't negligible as well.

I know that an ULT stack could be grabbed from xstream A cache but put to xstream B cache at the end, but will this cause any issue?

If the mmap'd stack is used for debugging purpose only, then I think the performance impact could be acceptable.

Reply:
Indeed, this not intended to be used in production as in any case using mmap as some performance cost.
From my side, I see two main usage:
- For support tasks on client side to detect which ULT is overusing is stack
- For CI to detect ULT stack overflow regression as we have experimented at LRZ with UCX library using discovery functions overflowing their ULT stacks.

The issue with having it per xstream is that you have to be sure that the xstream creating an ULT and the one deallocating its resources have to be the same.  Indeed, the ULT deallocation is respectively done by the xstream running the ULT thread and the xstream calling ABT_thread_free() function for an unnamed ULT and a named ULT.
From my understanding, this condition is not respected in the DAOS code.

Reply:
I see, it looks good to me if it's not used for production.